### PR TITLE
[CDF-22126] 3D Valhall

### PR DESCRIPTION
**Pull request by BergsethCognite**
# Description

Please describe the change you have made.

## Checklist

- [x] Tests added/updated.
- [x] Run Demo Job Locally.
- [x] Documentation updated. (Readme)
- [ ] Changelogs updated in [CHANGELOG.cdf-tk.md](https://github.com/cognitedata/toolkit/blob/main/CHANGELOG.cdf-tk.md).
- [x] Template changelogs updated in [CHANGELOG.templates.md](https://github.com/cognitedata/toolkit/blob/main/CHANGELOG.templates.md).
- [ ] Version bumped.
  [_version.py](https://github.com/cognitedata/toolkit/blob/main/cognite/cognite_toolkit/_version.py) and
  [pyproject.toml](https://github.com/cognitedata/toolkit/blob/main/pyproject.toml) per [semantic versioning](https://semver.org/).

### CODE REVIEW

File: cognite_toolkit/cognite_modules/examples/cdf_data_pipeline_3d_valhall/README.md [hunk 1]
```diff
@@ -0,0 +1,116 @@
+# cdf_data_pipeline_3d_valhall
+
+This module relies on these example data modules being loaded:
+
+- cdf_oid_example_data
+- cdf_data_pipeline_asset_valhall
+- The Valhall 3D model uploaded and processed by CDF 3D pipeline
+
+The module creates a simple data pipeline for contextualization of a 3D model (Valhall) with
+the asset data based on the OID example module. In addition the module contains and example
```
Comment (doctrino) [inline cognite_toolkit/cognite_modules/examples/cdf_data_pipeline_3d_valhall/README.md:10]:
```suggestion
the asset data based on the OID example module. In addition, the module contains and example
```

File: cognite_toolkit/cognite_modules/examples/cdf_data_pipeline_3d_valhall/README.md [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+File extractor configuration that can be used to upload the 3D model from your file system.
+
+Proccesing workflow for the contextualization are:
```
Comment (doctrino) [inline cognite_toolkit/cognite_modules/examples/cdf_data_pipeline_3d_valhall/README.md:13]:
```suggestion
Processing workflow for contextualization are:
```

File: cognite_toolkit/cognite_modules/examples/cdf_data_pipeline_3d_valhall/README.md [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+| Variable | Description |
+|----------|-------------|
+| location_name | The location for your data, name used in all resource type related to data pipeline. We use oid (Open Industrial Data) |
+| source_name | The name of the source making it possible to identify where the data originates from, ex: 'workmate', 'sap', 'oracle',..|
+| 3d_dataset | The name of data set used for time series in this example, must correspond to name used in example data|
+| 3d_model_name | The name of 3d name used in example data|
+| external_root_id_asset | The external ID for the Asset name used in example data|
+| 3d_location_extractor_group_source_id | Object/ Source ID for security group in IdP. Used to run integration/extractor|
+| 3d_location_processing_group_source_id | Object/ Source ID for security group in IdP. Used to run CDF processing/contextualization|
+| 3d_location_read_group_source_id | Object/ Source ID for security group in IdP. Used to read file data|
```
Comment (doctrino) [inline cognite_toolkit/cognite_modules/examples/cdf_data_pipeline_3d_valhall/README.md:104]:
```suggestion
| Variable                               | Description                                                                                                              |
|----------------------------------------|--------------------------------------------------------------------------------------------------------------------------|
| location_name                          | The location for your data, name used in all resource type related to data pipeline. We use oid (Open Industrial Data)   |
| source_name                            | The name of the source making it possible to identify where the data originates from, ex: 'workmate', 'sap', 'oracle',.. |
| 3d_dataset                             | The name of data set used for time series in this example, must correspond to name used in example data                  |
| 3d_model_name                          | The name of 3d name used in example data                                                                                 |
| external_root_id_asset                 | The external ID for the Asset name used in example data                                                                  |
| 3d_location_extractor_group_source_id  | Object/ Source ID for security group in IdP. Used to run integration/extractor                                           |
| 3d_location_processing_group_source_id | Object/ Source ID for security group in IdP. Used to run CDF processing/contextualization                                |
| 3d_location_read_group_source_id       | Object/ Source ID for security group in IdP. Used to read file data                                                      |
```

File: cognite_toolkit/cognite_modules/examples/cdf_data_pipeline_3d_valhall/README.md [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+For the test, the model was uploaded to CDF using the UI and the 3D pipeline in CDF.
+
+## Managed resources
+
+This module manages the following resources:
+
+1. auth:
+   - Name: `gp_3d_oid_extractor`
+     - Content: Authorization group used by the extractor writing data to CDF files and RAW,
+       extraction pipeline and update of extraction pipeline run
```
Comment (doctrino) [inline cognite_toolkit/cognite_modules/examples/cdf_data_pipeline_3d_valhall/README.md:31]:
```suggestion
       extraction pipeline and update of the extraction pipeline run
```

File: cognite_toolkit/cognite_modules/examples/cdf_data_pipeline_3d_valhall/README.md [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+   - db/table: `3d_oid_opcua/contextualization_good`
+     - Content: result from contextualization with mappings that was added to CDF between 3d and assets.
+       Table can be used for quality control and input to manual tuning
+   - db/table: `3d_oid_opcua/contextualization_bad`
+     - Content: result from contextualization with mappings NOT added to CDF between 3d and assets.
+       Table can be used for quality control and input to manual tuning
+   - db/table: `3d_oid_opcua/contextualization_manual_input`
+     - Content: manual input to contextualization process. Table can be used as:
+       - a migration table read existing mappings from old project and insert to this table
+         to re-map in new project
```
Comment (doctrino) [inline cognite_toolkit/cognite_modules/examples/cdf_data_pipeline_3d_valhall/README.md:74]:
```suggestion
     - Content: manual input to the contextualization process. Table can be used as:
       - a migration table reads existing mappings from the old project and insert to this table
         to re-map in the new project
```

File: cognite_toolkit/cognite_modules/examples/cdf_data_pipeline_3d_valhall/README.md [hunk 6]
```diff
@@ -0,0 +1,116 @@
...
+
+### Illustration of the time 3d data pipeline
+
+![image](https://github.com/cognitedata/toolkit/assets/31886431/f1129181-bab0-42cb-8366-860e8fb30d7e)
+
+### Illustration of the contextualization workflow
+
+With usage of the output stored in the good /bad table to process false positive or just manually map
+content not able to be processed automatically, you can create a contextualization workflow.
+Included in the workflow there could also be a rule module using properties of the data or external
```
Comment (doctrino) [inline cognite_toolkit/cognite_modules/examples/cdf_data_pipeline_3d_valhall/README.md:86]:
```suggestion
Included in the workflow, there could also be a rule module using properties of the data or external
```

File: cognite_toolkit/cognite_modules/examples/cdf_data_pipeline_3d_valhall/README.md [hunk 7]
```diff
@@ -0,0 +1,116 @@
...
+| 3d_model_name | The name of 3d name used in example data|
+| external_root_id_asset | The external ID for the Asset name used in example data|
+| 3d_location_extractor_group_source_id | Object/ Source ID for security group in IdP. Used to run integration/extractor|
+| 3d_location_processing_group_source_id | Object/ Source ID for security group in IdP. Used to run CDF processing/contextualization|
+| 3d_location_read_group_source_id | Object/ Source ID for security group in IdP. Used to read file data|
+
+## Usage
+
+You should copy and rename an example module into the `custom_modules` directory (remove any `cdf_` prefixes) and make
+your own modifications. You should then update the `selected_modules_and_packages:` section in your `config.[env].yaml`
```
Comment (doctrino) [inline cognite_toolkit/cognite_modules/examples/cdf_data_pipeline_3d_valhall/README.md:109]:
```suggestion
your own modifications. You should then update the `selected:` section in your `config.[env].yaml`
```

File: cognite_toolkit/cognite_modules/examples/cdf_data_pipeline_3d_valhall/data_sets/dataset.yaml [hunk 8]
```diff
@@ -0,0 +1,14 @@
+externalId: ds_3d_{{location_name}}
```
Comment (doctrino) [inline cognite_toolkit/cognite_modules/examples/cdf_data_pipeline_3d_valhall/data_sets/dataset.yaml:1]: This is duplicated of https://github.com/cognitedata/toolkit/blob/main/cognite_toolkit/cognite_modules/examples/cdf_oid_example_data/data_sets/data_sets.yaml#L10

Should remove if from the OID example?

**Review by doctrino — COMMENTED** [review]
Reviewed the README.md and building of module.

Will complete review once full 3D model is available.

File: cognite_toolkit/cognite_modules/examples/cdf_data_pipeline_3d_valhall/data_sets/dataset.yaml (hunk 8 above)
Comment (BergsethCognite) [reply cognite_toolkit/cognite_modules/examples/cdf_data_pipeline_3d_valhall/data_sets/dataset.yaml:1]: This data set contains metadata about the source and used raw tables, so don't think we should remove.

**Review by BergsethCognite — COMMENTED** [review]
@doctrino Test now fails with :  Exceeded maximum number of containers per project. The maximum number of containers for your project is 1000

3D model is shared with you on Google drive

File: cognite_toolkit/_cdf_tk/prototypes/resource_loaders.py [hunk 9]
```diff
@@ -181,6 +183,7 @@ class ThreeDModelLoader(ResourceLoader[str, ThreeDModelWrite, ThreeDModel, Three
     kind = "3DModel"
     dependencies = frozenset({DataSetsLoader})
     _doc_url = "3D-Models/operation/create3DModels"
+    item_name = "Revision"
```
Comment (doctrino) [inline cognite_toolkit/_cdf_tk/prototypes/resource_loaders.py:186]:
```suggestion
    item_name = "revisions"
```

File: cognite_toolkit/cognite_modules/examples/cdf_data_pipeline_3d_valhall/3dmodels/valhall.3DModel.yaml [hunk 10]
```diff
@@ -0,0 +1,2 @@
+name: {{3d_model_name}}
+dataSetExternalId: {{3d_dataset}}
```
Comment (doctrino) [inline cognite_toolkit/cognite_modules/examples/cdf_data_pipeline_3d_valhall/3dmodels/valhall.3DModel.yaml:2]: Added 3D Model explicitly

File: cognite_toolkit/cognite_modules/examples/cdf_data_pipeline_3d_valhall/functions/functions.yaml [hunk 11]
```diff
@@ -10,8 +10,6 @@
     # The below two environment variables are set by the Toolkit
     ENV_TYPE: '${CDF_BUILD_TYPE}'
     CDF_ENV: '${CDF_ENVIRON}'
-  # Number of cores, not available in Azure
-  cpu: 0.60
```
Comment (doctrino) [inline cognite_toolkit/cognite_modules/examples/cdf_data_pipeline_3d_valhall/functions/functions.yaml]: Instead let the server set the default based on whether it is deployed on Azuer/AWS/Google

File: cognite_toolkit/cognite_modules/examples/cdf_data_pipeline_3d_valhall/functions/schedules.yaml [hunk 12]
```diff
@@ -3,7 +3,7 @@
   description: "Run every 30 minute"
   cronExpression: "15 * * * *"
   data:
-    ExtractionPipelineExtId: 'ep_ctx_3d_{{location_name}}_{{source_name}}_asset'
+    ExtractionPipelineExtId: 'ep_ctx_3d_{{location_name}}_{{source_name}}_annotation'
```
Comment (doctrino) [inline cognite_toolkit/cognite_modules/examples/cdf_data_pipeline_3d_valhall/functions/schedules.yaml:6]: This pipeline does not exists, assume it is should be this one.

**Review by doctrino — APPROVED** [review]

**Pull request merged**